# **Necessary Imports**

In [252]:
# Import required libraries
import numpy as np
import pandas as pd
import sqlite3 as sql

# **Notes:**

* Use pandas dataframe (df) to load the data.

* Use numpy or pandas operations for the requested tasks unless otherwise specified.

* Use of naive for loops is not considered valid.

* For the db operations, use sqlite3 library in Python

# **PART- 1**

**Set of Questions on World Happiness Report Dataset (40 Pts)**

**Dataset Path: "Data/WorldHappinessReport.csv"**

# Q1. Basic Data Exploration **(10 Pts)**

1. Load the data into Pandas DataFrame and print the column names. **(1 pts)**
2. Display the first and last 10 rows of the dataset. **(1 pts)**
3. Display random 10 rows. **(1 pts)**
4. Display random 10 rows with a seed, ie, it outputs the same random rows everytime it is run. **(1 pts)**
5. Display every 10th row. **(1 pts)**
6. Get the summary stats of the dataset regarding each column. **(2 pts)**
7. Identify and count the number of missing values in each column. **(3 pts)**

In [254]:
#1 Load the data into Pandas DataFrame and print the column names
df = pd.read_csv("Data/WorldHappinessReport.csv")
print(df.columns)

Index(['Overall rank', 'Country or region', 'Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption'],
      dtype='object')


In [256]:
#2 Display the first and last 10 rows of the dataset.
df.head(10) # Shows the first 10 rows

Overall rank Country or region  Score  GDP per capita  Social support  \
0             1           Finland  7.632           1.305           1.592   
1             2            Norway  7.594           1.456           1.582   
2             3           Denmark  7.555           1.351           1.590   
3             4           Iceland  7.495           1.343           1.644   
4             5       Switzerland  7.487           1.420           1.549   
5             6       Netherlands  7.441           1.361           1.488   
6             7            Canada  7.328           1.330           1.532   
7             8       New Zealand  7.324           1.268           1.601   
8             9            Sweden  7.314           1.355           1.501   
9            10         Australia  7.272           1.340           1.573   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
0                    0.874                         0.681       0.202   
1                    0.861                         0.686       0.286   
2                    0.868                         0.683       0.284   
3                    0.914                         0.677       0.353   
4                    0.927                         0.660       0.256   
5                    0.878                         0.638       0.333   
6                    0.896                         0.653       0.321   
7                    0.876                         0.669       0.365   
8                    0.913                         0.659       0.285   
9                    0.910                         0.647       0.361   

   Perceptions of corruption  
0                      0.393  
1                      0.340  
2                      0.408  
3                      0.138  
4                      0.357  
5                      0.295  
6                      0.291  
7                      0.389  
8                      0.383  
9                      0.302

In [259]:
df.tail(10) # Shows the last 10 rows

Overall rank         Country or region  Score  GDP per capita  \
146           147                    Malawi  3.587           0.186   
147           148                     Haiti  3.582           0.315   
148           149                   Liberia  3.495           0.076   
149           150                     Syria  3.462           0.689   
150           151                    Rwanda  3.408           0.332   
151           152                     Yemen  3.355           0.442   
152           153                  Tanzania  3.303           0.455   
153           154               South Sudan  3.254           0.337   
154           155  Central African Republic  3.083           0.024   
155           156                   Burundi  2.905           0.091   

     Social support  Healthy life expectancy  Freedom to make life choices  \
146           0.541                    0.306                         0.531   
147           0.714                    0.289                         0.025   
148           0.858                    0.267                         0.419   
149           0.382                    0.539                         0.088   
150           0.896                    0.400                         0.636   
151           1.073                    0.343                         0.244   
152           0.991                    0.381                         0.481   
153           0.608                    0.177                         0.112   
154           0.000                    0.010                         0.305   
155           0.627                    0.145                         0.065   

     Generosity  Perceptions of corruption  
146       0.210                      0.080  
147       0.392                      0.104  
148       0.206                      0.030  
149       0.376                      0.144  
150       0.200                      0.444  
151       0.083                      0.064  
152       0.270                      0.097  
153       0.224                      0.106  
154       0.218                      0.038  
155       0.149                      0.076

In [258]:
#3 Display the random 10 rows
df.sample(n=10)

Overall rank        Country or region  Score  GDP per capita  \
129           130                  Myanmar  4.308           0.682   
153           154              South Sudan  3.254           0.337   
133           134                    Niger  4.166           0.131   
143           144                 Zimbabwe  3.692           0.357   
50             51                 Slovenia  5.948           1.219   
103           104  Palestinian Territories  4.743           0.642   
101           102                Venezuela  4.806           0.996   
54             55                Mauritius  5.891           1.090   
94             95                  Vietnam  5.103           0.715   
115           116                Sri Lanka  4.471           0.918   

     Social support  Healthy life expectancy  Freedom to make life choices  \
129           1.174                    0.429                         0.580   
153           0.608                    0.177                         0.112   
133           0.867                    0.221                         0.390   
143           1.094                    0.248                         0.406   
50            1.506                    0.856                         0.633   
103           1.217                    0.602                         0.266   
101           1.469                    0.657                         0.133   
54            1.387                    0.684                         0.584   
94            1.365                    0.702                         0.618   
115           1.314                    0.672                         0.585   

     Generosity  Perceptions of corruption  
129       0.598                      0.178  
153       0.224                      0.106  
133       0.175                      0.099  
143       0.132                      0.099  
50        0.160                      0.051  
103       0.086                      0.076  
101       0.056                      0.052  
54        0.245                      0.050  
94        0.177                      0.079  
115       0.307                      0.050

In [260]:
#4 sample using seed
df.sample(n=10, random_state=20) # Shows 10 randomly selected rows with a fixed random seed (20)

Overall rank Country or region  Score  GDP per capita  Social support  \
114           115        Bangladesh  4.500           0.532           0.850   
51             52           Romania  5.945           1.116           1.219   
76             77          Portugal  5.410           1.188           1.429   
133           134             Niger  4.166           0.131           0.867   
72             73           Belarus  5.483           1.039           1.498   
108           109           Senegal  4.631           0.429           1.117   
122           123        Mozambique  4.417           0.198           0.902   
21             22             Malta  6.627           1.270           1.525   
86             87        Azerbaijan  5.201           1.024           1.161   
12             13        Costa Rica  7.072           1.010           1.459   

     Healthy life expectancy  Freedom to make life choices  Generosity  \
114                    0.579                         0.580       0.153   
51                     0.726                         0.528       0.088   
76                     0.884                         0.562       0.055   
133                    0.221                         0.390       0.175   
72                     0.700                         0.307       0.101   
108                    0.433                         0.406       0.138   
122                    0.173                         0.531       0.206   
21                     0.884                         0.645       0.376   
86                     0.603                         0.430       0.031   
12                     0.817                         0.632       0.143   

     Perceptions of corruption  
114                      0.144  
51                       0.001  
76                       0.017  
133                      0.099  
72                       0.154  
108                      0.082  
122                      0.158  
21                       0.142  
86                       0.176  
12                       0.101

In [261]:
#5 Display every 10th row
df.iloc[::10]

Overall rank Country or region  Score  GDP per capita  Social support  \
0               1           Finland  7.632           1.305           1.592   
10             11    United Kingdom  7.190           1.244           1.433   
20             21    Czech Republic  6.711           1.233           1.489   
30             31           Uruguay  6.379           1.093           1.459   
40             41         Nicaragua  6.141           0.668           1.319   
50             51          Slovenia  5.948           1.219           1.506   
60             61            Cyprus  5.762           1.229           1.191   
70             71       Philippines  5.524           0.775           1.312   
80             81        Montenegro  5.347           1.017           1.279   
90             91           Nigeria  5.155           0.689           1.172   
100           101             Nepal  4.880           0.425           1.228   
110           111           Tunisia  4.592           0.900           0.906   
120           121      Burkina Faso  4.424           0.314           1.097   
130           131              Chad  4.301           0.358           0.907   
140           141           Lesotho  3.808           0.472           1.215   
150           151            Rwanda  3.408           0.332           0.896   

     Healthy life expectancy  Freedom to make life choices  Generosity  \
0                      0.874                         0.681       0.202   
10                     0.888                         0.464       0.262   
20                     0.854                         0.543       0.064   
30                     0.771                         0.625       0.130   
40                     0.700                         0.527       0.208   
50                     0.856                         0.633       0.160   
60                     0.909                         0.423       0.202   
70                     0.513                         0.643       0.120   
80                     0.729                         0.259       0.111   
90                     0.048                         0.462       0.201   
100                    0.539                         0.526       0.302   
110                    0.690                         0.271       0.040   
120                    0.254                         0.312       0.175   
130                    0.053                         0.189       0.181   
140                    0.079                         0.423       0.116   
150                    0.400                         0.636       0.200   

     Perceptions of corruption  
0                        0.393  
10                       0.082  
20                       0.034  
30                       0.155  
40                       0.128  
50                       0.051  
60                       0.035  
70                       0.105  
80                       0.081  
90                       0.032  
100                      0.078  
110                      0.063  
120                      0.128  
130                      0.060  
140                      0.112  
150                      0.444

In [262]:
#6 Get the summary stats of the dataset regarding each column
df.describe() # Outputs key statistics such as mean, median, and standard deviation for numeric columns

Overall rank       Score  GDP per capita  Social support  \
count    156.000000  156.000000      156.000000      156.000000   
mean      78.500000    5.375917        0.891449        1.213237   
std       45.177428    1.119506        0.391921        0.302372   
min        1.000000    2.905000        0.000000        0.000000   
25%       39.750000    4.453750        0.616250        1.066750   
50%       78.500000    5.378000        0.949500        1.255000   
75%      117.250000    6.168500        1.197750        1.463000   
max      156.000000    7.632000        2.096000        1.644000   

       Healthy life expectancy  Freedom to make life choices  Generosity  \
count               156.000000                    156.000000  156.000000   
mean                  0.597346                      0.454506    0.181006   
std                   0.247579                      0.162424    0.098471   
min                   0.000000                      0.000000    0.000000   
25%                   0.422250                      0.356000    0.109500   
50%                   0.644000                      0.487000    0.174000   
75%                   0.777250                      0.578500    0.239000   
max                   1.030000                      0.724000    0.598000   

       Perceptions of corruption  
count                 155.000000  
mean                    0.112000  
std                     0.096492  
min                     0.000000  
25%                     0.051000  
50%                     0.082000  
75%                     0.137000  
max                     0.457000

In [264]:
#7 Identify and count the number of missing values in each column
print(df.isnull().sum(axis=0)) # Displays the total number of null values for each column

Overall rank                    0
Country or region               0
Score                           0
GDP per capita                  0
Social support                  0
Healthy life expectancy         0
Freedom to make life choices    0
Generosity                      0
Perceptions of corruption       1
dtype: int64


# Q2. Basic Data Statistics: **(10 Pts)**

1. Calculate the mean, median, and standard deviation of the happiness Score across all countries. **(1 Pts)**
2. Identify the top 5 happiest countries and the bottom 5 least happy countries based on happiness Score. **(1 Pts)**
3. Find the country with the highest Healthy Life Expectancy. What is the corresponding happiness Score? **(2 Pts)**
4. Calculate the average Corruption Perceptions for countries with a happiness Score greater than 6. **(2 Pts)**
5. Are there any countries where Generosity is higher than Social Support? List them. **(2 Pts)**
6. What is the range (max-min) of GDP per Capita values in the dataset? **(2 Pts)**

In [185]:
#1  
print(df["Score"].mean()) # Outputs the mean value of the 'Score' column

5.375916666666667


In [186]:
print(df["Score"].median()) # Displays the median value of the 'Score' column

5.378


In [187]:
print(df["Score"].std()) # Displays the standard deviation of the 'Score' column

1.119505852731943


In [266]:
#2 
# Sort the DataFrame by the 'Score' column in descending order
sorted_df = df.sort_values(by="Score", ascending=False, inplace=False)

# Display the top countries based on score
print(sorted_df.head()["Country or region"])

# Display the bottom countries based on score
print(sorted_df.tail()["Country or region"])

0        Finland
1         Norway
2        Denmark
3        Iceland
4    Switzerland
Name: Country or region, dtype: object
151                       Yemen
152                    Tanzania
153                 South Sudan
154    Central African Republic
155                     Burundi
Name: Country or region, dtype: object


In [189]:
#3
# Get the index of the country with the highest healthy life expectancy
country_id = df["Healthy life expectancy"].idxmax()

# Display the country with the highest healthy life expectancy
print(df.loc[country_id]["Country or region"])

Hong Kong


In [190]:
# Display the score of the country with the highest healthy life expectancy
print(df.loc[country_id]["Score"])

5.43


In [191]:
#4
# Create a new DataFrame with countries having a score greater than 6
new_df = df.loc[df["Score"]>6]

# Calculate the average perceptions of corruption for the new_df
average = new_df["Perceptions of corruption"].mean()

# Display the average perceptions of corruption
print(average)

0.1782888888888889


In [192]:
#5
# Create a new DataFrame for countries where generosity is greater than social support
new_df = df.loc[df["Generosity"] > df["Social support"]]
# List the countries where the level of generosity is greater than the level of social support
print(new_df["Country or region"])

154    Central African Republic
Name: Country or region, dtype: object


In [193]:
#6
# Calculate the range of GDP per capita by subtracting the minimum from the maximum
range = df["GDP per capita"].max() - df["GDP per capita"].min()

# Display the range of GDP per capita
print(range)

2.096


# Q3. Data Filtering **(20 Pts)**

1. How many countries have a Happiness Score greater than 7.5? List their names.**(2 Pts)**
2. List all countries where the Corruption Perception score is below 0.2. **(2 Pts)**
3. Filter and list the top 5 countries by Generosity. What is their average Happiness Score? **(2 Pts)**
4. Find countries where GDP per Capita is less than 0.5 but the Happiness Score is above 4. List the countries. **(2 Pts)**
5. Filter countries where Social Support is higher than Freedom to Make Life **(2 Pts)**
6. Find countries in the top 10% of Healthy Life Expectancy and list their names. **(2 Pts)**
7. Filter and list countries where Generosity is less than Social Support. What is the average Happiness Score of these countries? **(2 Pts)**
8.  Find all countries with a Happiness Score below the median happiness score. How many are there, and what are the common characteristics of these countries? **(3 Pts)**
9. Filter countries where both Freedom to Make Life Choices and Healthy Life Expectancy are above the global median. How many such countries are there? **(3 Pts)**

In [194]:
#1
#Create a new DataFrame from the values in the original DataFrame where the score is above 7.5.
new_df = df.loc[df["Score"] > 7.5] 

#It displays the number of countries that meet this condition.
print(len(new_df))

3


In [195]:
#It displays the names of the countries that meet this condition.
print(new_df["Country or region"])

0    Finland
1     Norway
2    Denmark
Name: Country or region, dtype: object


In [268]:
#2
#Create a new DataFrame with the entries where the Corruption Perception score is below 0.2.
new_df = df.loc[df["Perceptions of corruption"] < 0.2]

#It displays the names of the countries in the new_df.
print(new_df["Country or region"])

3                       Iceland
10               United Kingdom
12                   Costa Rica
17                United States
20               Czech Republic
                 ...           
151                       Yemen
152                    Tanzania
153                 South Sudan
154    Central African Republic
155                     Burundi
Name: Country or region, Length: 134, dtype: object


In [270]:
#3
# Sort the DataFrame by the 'Generosity' column in descending order
sorted_df = df.sort_values(by="Generosity", ascending=False, inplace=False)

#List the top 5 countries by generosity.
print(sorted_df.head()["Country or region"])

129      Myanmar
95     Indonesia
147        Haiti
21         Malta
149        Syria
Name: Country or region, dtype: object


In [198]:
# Calculate the average Happiness Score of the top 5 countries
print(sorted_df.head()["Score"].mean())

4.6144


In [271]:
#4
# Create a new DataFrame with countries having GDP per capita less than 0.5 and Score greater than 4
new_df = df.loc[(df["GDP per capita"] < 0.5) & (df["Score"] > 4)]

# Display the names of the countries in the new DataFrame
new_df["Country or region"]

87           Tajikistan
97              Somalia
100               Nepal
108             Senegal
112        Sierra Leone
117                Mali
120        Burkina Faso
122          Mozambique
123               Kenya
126            Ethiopia
130                Chad
131    Congo (Kinshasa)
133               Niger
134              Uganda
135               Benin
Name: Country or region, dtype: object

In [200]:
#5
# Create a new DataFrame with countries where social support is greater than freedom to make life choices
new_df = df.loc[df["Social support"] > df["Freedom to make life choices"]]

# Display the new DataFrame
new_df

Overall rank Country or region  Score  GDP per capita  Social support  \
0               1           Finland  7.632           1.305           1.592   
1               2            Norway  7.594           1.456           1.582   
2               3           Denmark  7.555           1.351           1.590   
3               4           Iceland  7.495           1.343           1.644   
4               5       Switzerland  7.487           1.420           1.549   
..            ...               ...    ...             ...             ...   
150           151            Rwanda  3.408           0.332           0.896   
151           152             Yemen  3.355           0.442           1.073   
152           153          Tanzania  3.303           0.455           0.991   
153           154       South Sudan  3.254           0.337           0.608   
155           156           Burundi  2.905           0.091           0.627   

     Healthy life expectancy  Freedom to make life choices  Generosity  \
0                      0.874                         0.681       0.202   
1                      0.861                         0.686       0.286   
2                      0.868                         0.683       0.284   
3                      0.914                         0.677       0.353   
4                      0.927                         0.660       0.256   
..                       ...                           ...         ...   
150                    0.400                         0.636       0.200   
151                    0.343                         0.244       0.083   
152                    0.381                         0.481       0.270   
153                    0.177                         0.112       0.224   
155                    0.145                         0.065       0.149   

     Perceptions of corruption  
0                        0.393  
1                        0.340  
2                        0.408  
3                        0.138  
4                        0.357  
..                         ...  
150                      0.444  
151                      0.064  
152                      0.097  
153                      0.106  
155                      0.076  

[154 rows x 9 columns]

In [201]:
#6
# Calculate the 90th percentile value for Healthy Life Expectancy
value = df["Healthy life expectancy"].quantile(0.90)

# Create a new DataFrame with countries in the top 10% of Healthy Life Expectancy
new_df = df.loc[df["Healthy life expectancy"] >= value]

# Display the new DataFrame
new_df["Country or region"]

3             Iceland
4         Switzerland
6              Canada
8              Sweden
9           Australia
15            Belgium
16         Luxembourg
22             France
33          Singapore
35              Spain
46              Italy
53              Japan
56        South Korea
57    Northern Cyprus
60             Cyprus
75          Hong Kong
Name: Country or region, dtype: object

In [202]:
#7
#Create a new DataFrame with the entries where the generosity value is less than the social support value.
new_df = df.loc[df["Generosity"] < df["Social support"]]

# Display the new DataFrame
new_df["Country or region"]

0          Finland
1           Norway
2          Denmark
3          Iceland
4      Switzerland
          ...     
150         Rwanda
151          Yemen
152       Tanzania
153    South Sudan
155        Burundi
Name: Country or region, Length: 155, dtype: object

In [203]:
# # Display the  average of the score values in the new DataFrame
average = new_df["Score"].mean()
print(average)

5.390709677419355


In [204]:
#8
# Calculate the median happiness score
medianScore = df["Score"].median()

# Create a new DataFrame with countries having a score less than the median score
new_df = df.loc[df["Score"] < medianScore]

# Display the number of countries in the new DataFrame
print(len(new_df))

78


In [205]:
# Display the common characteristics of countries with a Happiness Score below the median
new_df.describe()

Overall rank      Score  GDP per capita  Social support  \
count     78.000000  78.000000       78.000000       78.000000   
mean     117.500000   4.449321        0.621256        1.019423   
std       22.660538   0.623645        0.308177        0.288077   
min       79.000000   2.905000        0.000000        0.000000   
25%       98.250000   4.139500        0.357250        0.860250   
50%      117.500000   4.451500        0.623500        1.083000   
75%      136.750000   4.980250        0.910500        1.215000   
max      156.000000   5.358000        1.154000        1.517000   

       Healthy life expectancy  Freedom to make life choices  Generosity  \
count                78.000000                     78.000000   78.000000   
mean                  0.432962                      0.386885    0.176705   
std                   0.221797                      0.169637    0.105170   
min                   0.000000                      0.000000    0.000000   
25%                   0.254250                      0.283000    0.112250   
50%                   0.423000                      0.419000    0.174000   
75%                   0.612750                      0.515500    0.219500   
max                   0.879000                      0.696000    0.598000   

       Perceptions of corruption  
count                  78.000000  
mean                    0.088295  
std                     0.064311  
min                     0.000000  
25%                     0.052000  
50%                     0.077500  
75%                     0.105500  
max                     0.444000

In [206]:
#9Filter countries where both Freedom to Make Life Choices and Healthy Life Expectancy are above the global median. How many such countries are there? 
# Calculate the median for Freedom to Make Life Choices and Healthy Life Expectancy
freedomMedian = df["Freedom to make life choices"].median()
healtyMedian = df ["Healthy life expectancy"].median()

# Filter the DataFrame for countries where both metrics are above the global median
new_df = df.loc[(df["Freedom to make life choices"] > freedomMedian) & (df ["Healthy life expectancy"] > healtyMedian)]

# Display the number of countries that meet the criteria
print(len(new_df))

47


# **PART- 2**

**Set of Questions on the passenger_demographics.csv and passenger_survival.csv** **(25 Pts)**

**Dataset Path: "Data/passenger_demographics.csv" and "Data/passenger_survival.csv"**

# Q1. Data Merging and Consistency Checking


1. Merge the datasets on PassengerId into a single DataFrame. How many rows are there in the merged dataset? **(3 Pts)**
2. Perform a consistency check for duplicated rows. After merging, check for any duplicated rows in the merged dataset and provide a summary of your findings. Provide the details (average, mean, etc.) of each feature in the merged DataFrame. **(3 Pts)**
3. Determine the survival rate for each passenger class and gender. Analyze how the survival rate varies among different classes (Pclass) and genders. **(3 Pts)**
4. Analyze the effect of age on survival probability. Determine whether age significantly affects the probability of survival. Use statistical techniques to explore this relationship.(use binning) **(4 Pts)**
5. Investigate the impact of family size on survival. Determine if family size (based on the number of siblings/spouses and parents/children aboard) has any correlation with survival. **(4 Pts)**
6. Determine the most common embarkation point among survivors. Find out which embarkation point (C, Q, or S) had the highest survival rate. **(4 Pts)**
7. Evaluate the correlation between Fare and Survival. Determine if there is a correlation between the fare paid and the survival status of the passengers. **(4 Pts)**

In [207]:
df1 = pd.read_csv("Data/passenger_demographics.csv") # Load the passenger demographics dataset from the specified CSV file into a DataFrame

In [208]:
df2 = pd.read_csv("Data/passenger_survival.csv") # Load the passenger survival dataset from the specified CSV file into a DataFrame

In [209]:
#1 
# Merge the two DataFrames on 'PassengerId' using an outer join
merged_df  = pd.merge(df1, df2, on='PassengerId', how='outer')

# Display the number of rows in the merged DataFrame
merged_df.shape[0]

891

In [210]:
#2
# Display the duplicate rows in the merged DataFrame
print(merged_df[merged_df.duplicated()]) 

Empty DataFrame
Columns: [PassengerId, Pclass, Sex, Age, SibSp, Parch, Survived, Fare, Embarked]
Index: []


In [211]:
# Provides summary statistics for the numeric columns in the merged DataFrame
merged_df.describe()

PassengerId      Pclass         Age       SibSp       Parch  \
count   891.000000  891.000000  714.000000  891.000000  891.000000   
mean    446.000000    2.308642   29.699118    0.523008    0.381594   
std     257.353842    0.836071   14.526497    1.102743    0.806057   
min       1.000000    1.000000    0.420000    0.000000    0.000000   
25%     223.500000    2.000000   20.125000    0.000000    0.000000   
50%     446.000000    3.000000   28.000000    0.000000    0.000000   
75%     668.500000    3.000000   38.000000    1.000000    0.000000   
max     891.000000    3.000000   80.000000    8.000000    6.000000   

         Survived        Fare  
count  891.000000  891.000000  
mean     0.383838   32.204208  
std      0.486592   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      1.000000   31.000000  
max      1.000000  512.329200

In [212]:
#3
# Calculate and display the mean survival rate grouped by passenger class (Pclass)
print(merged_df.groupby('Pclass')['Survived'].mean())

Pclass
1    0.629630
2    0.472826
3    0.242363
Name: Survived, dtype: float64


When the data is analyzed, the survival rate is highest in the 1st passenger class and lowest in the 3rd passenger class.

In [213]:
# Calculate and display the mean survival rate grouped by gender
print(merged_df.groupby('Sex')['Survived'].mean())

Sex
female    0.742038
male      0.188908
Name: Survived, dtype: float64


When the data is analyzed, the survival rate of females is higher compared to males.

In [214]:
#4
print(merged_df['Age'].max()) # Calculate the maximum age to segment the ranges.

80.0


In [273]:
# Define bins and labels for age intervals
bins = [0,10,20,30,40,50,60,70,80]
labels = ['0-10', '11-20','21-30','31-40','41-50','51-60','61-70','71-80']

# Create a new column 'AgeInterval' in the merged DataFrame to categorize ages into defined intervals
merged_df['AgeInterval'] = pd.cut(merged_df['Age'], bins= bins, labels=labels)

# Calculate and display the mean survival rate for each age interval

print(merged_df.groupby('AgeInterval')['Survived'].mean())

AgeInterval
0-10     0.593750
11-20    0.382609
21-30    0.365217
31-40    0.445161
41-50    0.383721
51-60    0.404762
61-70    0.235294
71-80    0.200000
Name: Survived, dtype: float64


C:\Users\fatma\AppData\Local\Temp\ipykernel_9612\1838202679.py:10: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  print(merged_df.groupby('AgeInterval')['Survived'].mean())


In [216]:
#5
# Calculate the family size by adding the number of siblings/spouses (SibSp) and parents/children (Parch) and adding 1 (the individual themselves)
merged_df['FamilySize'] = merged_df['SibSp'] + merged_df['Parch'] + 1 #

#Calculate the average of the survived values based on FamilySize.
merged_df.groupby('FamilySize')['Survived'].mean()

FamilySize
1     0.303538
2     0.552795
3     0.578431
4     0.724138
5     0.200000
6     0.136364
7     0.333333
8     0.000000
11    0.000000
Name: Survived, dtype: float64

When the data is examined, it is observed that smaller family sizes are associated with higher survival rates, while larger family sizes are linked to significantly lower survival rates.

In [217]:
#6
#Create a new DataFrame by selecting the survivors.
survivors_df = merged_df.loc[merged_df['Survived'] == 1]

#Calculate how many of each embarked number there are.
embarked_numbers = survivors_df['Embarked'].value_counts()

#Print the embarkation point with the highest embarked number.
print(embarked_numbers.idxmax())

S


In [218]:
#7
#To evaluate the correlation between Fare and Survived, it shows the average survival rates for specific fare ranges.
bins = [0, 50, 100, 150, 200, 300, 512.33]  
labels = ['0-50', '51-100', '101-150', '151-200', '201-300', '301+']
merged_df['FareInterval'] = pd.cut(merged_df['Fare'], bins= bins, labels=labels)
average_survival = merged_df.groupby('FareInterval')['Survived'].mean()
average_survival

C:\Users\fatma\AppData\Local\Temp\ipykernel_9612\1114208077.py:6: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  average_survival = merged_df.groupby('FareInterval')['Survived'].mean()


FareInterval
0-50       0.324022
51-100     0.654206
101-150    0.791667
151-200    0.666667
201-300    0.647059
301+       1.000000
Name: Survived, dtype: float64

When the data is evaluated, passengers who paid higher fares generally have a higher survival rate.

In [219]:
print(merged_df[['Fare', 'Survived']].corr()) # Calculate and display the correlation between Fare and Survived

              Fare  Survived
Fare      1.000000  0.257307
Survived  0.257307  1.000000


The value of 0.257307 is close to 0, indicating a weak positive relationship between Fare and Survived.

# **PART- 3**

**Set of Questions on SQL - using sqlite3 library!**

**Dataset Path: "Data/supermarket_sales.csv"**


The Supermarket Sales Dataset is a comprehensive dataset designed to provide insights into the sales performance of a supermarket.

We want to store this data in a relational database, with the following table schemas:

# Tables

## 1. Branches Table

| Column   | Type  | Key      | Description                              |
|----------|-------|----------|------------------------------------------|
| BranchID | INTEGER  | PRIMARY  | Unique identifier for each branch        |
| Branch | TEXT  |            |    Name of the supermarket branch        |
| City     | TEXT  |          | City where the branch is located         |



## 2. Products Table

| Column      | Type  | Key      | Description                               |
|-------------|-------|----------|-------------------------------------------|
| ProductID | INTEGER  | PRIMARY  | Unique identifier for each product line   |
| ProductLine | TEXT  |          | Category of the product |



## 3. Orders Table

| Column        | Type  | Key      | Description                                    |
|---------------|-------|----------|------------------------------------------------|
| OrderID       | TEXT  | PRIMARY  | Unique identifier for each order               |
| BranchID      | INTEGER  |   FK       | Identifier for the branch of the order    |
| ProductID   | INTEGER  |      FK    | Identifier linking to Products table ordered                      |
| Quantity      | INTEGER |        | Number of units ordered                        |
| UnitPrice   | REAL  |          | Price per unit of the product             |Total         | REAL  |          | Total sales value for the order               |
| OrderDate     | TEXT  |          | Date when the order was placed                |
| PaymentMethod | TEXT  |          | Method used for payment                       |
| CustomerType  | TEXT  |          | Type of customer (e.g., Member, Normal)     |
| Gender        | TEXT  |          | Gender of the customer                        |



Summary of Table Structures

* Branches Table contains information about different supermarket branches.

* Products Table details the various products available for purchase.
* Orders Table records the transactions made, linking products and branches to specific purchases.


# **Questions** **(35 Pts)**

1. Read the csv file and create this DB **(10 Pts)**
2. Write a query to find the top 3 product lines with the highest total quantity sold, using a LIMIT statement. **(4 Pts)**
3. Find the total revenue generated for each branch by product line **(4 Pts)**
4. Find the total sales for each branch **(3 Pts)**
5. Determine the number of orders placed by each gender **(3 Pts)**
6. List the product lines sold in each branch and the number of orders **(4 Pts)**
7. Find the highest quantity sold for a single order and the corresponding product line **(4 Pts)**
8. Check for duplicated invoices **(3 Pts)**


In [220]:
#1
df = pd.read_csv('Data/supermarket_sales.csv') # Read the Data/supermarket_sales.csv file

In [221]:
conn = sql.connect('sales.db') # Connect to the SQLite database named 'sales.db'

In [222]:
cursor = conn.cursor() # Create a cursor object to interact with the database

In [223]:
# Branches tablosunu sil
cursor.execute('DROP TABLE IF EXISTS BRANCHES')
conn.commit()

# Products tablosunu sil
cursor.execute('DROP TABLE IF EXISTS PRODUCTS')
conn.commit()

# Orders tablosunu sil
cursor.execute('DROP TABLE IF EXISTS ORDERS')
conn.commit()

# Diğer tabloları da silmek istersen, benzer şekilde devam edebilirsin:
# cursor.execute('DROP TABLE IF EXISTS TABLENAME')


In [224]:
# Create a table named 'BRANCHES' if it does not already exist, with BranchID as the primary key starting from 1 and set to autoincrement.
cursor.execute('''
CREATE TABLE IF NOT EXISTS BRANCHES ( 
    BranchID INTEGER PRIMARY KEY AUTOINCREMENT,
    Branch TEXT,
    City TEXT
)
''') 

In [225]:
# Create a table named 'PRODUCTS' if it does not already exist, with ProductID as the primary key
cursor.execute('''CREATE TABLE IF NOT EXISTS PRODUCTS (
    ProductID INTEGER PRIMARY KEY,
    ProductLine TEXT
)''')

In [226]:
## Create a table named 'ORDERS' if it does not already exist, with OrderID as the primary key
# BranchID and ProductID are foreign keys referencing the BRANCHES and PRODUCTS tables, respectively.
cursor.execute('''
CREATE TABLE IF NOT EXISTS ORDERS (
    OrderID TEXT PRIMARY KEY,
    BranchID INTEGER,
    ProductID INTEGER,
    Quantity INTEGER,
    UnitPrice REAL,
    OrderDate TEXT,
    PaymentMethod TEXT,
    CustomerType TEXT,
    Gender TEXT,
    FOREIGN KEY (BranchID) REFERENCES BRANCHES (BranchID),
    FOREIGN KEY (ProductID) REFERENCES PRODUCTS (ProductID)
)
''')

In [227]:
# Extract unique branches and cities from the DataFrame and store them in a new DataFrame
branches_df = df[['Branch', 'City']].drop_duplicates() 

# Then, append the unique values to the 'BRANCHES' table in the database.
branches_df.to_sql('BRANCHES', conn, if_exists='append', index=False)

3

In [228]:
# Rename the column 'Product line' to 'ProductLine' for consistency
df.rename(columns={'Product line': 'ProductLine'}, inplace=True) 

# Then, extract unique product lines from the DataFrame and store them in a new DataFrame
products_df = df[['ProductLine']].drop_duplicates()

# Append the unique product lines to the 'PRODUCTS' table in the database.
products_df.to_sql('PRODUCTS', conn, if_exists='append', index=False)

6

In [229]:
# Rename the columns for consistency
df.rename(columns={'Invoice ID': 'OrderID', 'Unit price': 'UnitPrice', 'Date':'OrderDate', 'Payment':'PaymentMethod', 'Customer type':'CustomerType'}, inplace=True)


In [230]:
# Create lists to hold the values forBranchID and ProductID
branch_ids = []
product_ids = []

for index, row in df.iterrows():
    # Query the BRANCHES table to get the BranchID for the current row's Branch
    cursor.execute("SELECT BranchID FROM BRANCHES WHERE Branch = ?", (row['Branch'],))
    branch_id = cursor.fetchone()[0] # Fetch the result
    branch_ids.append(branch_id) # Append the BranchID to branch_ids list.

    # Query the PRODUCTS table to get the ProductID for the current row's ProductLine
    cursor.execute("SELECT ProductID FROM PRODUCTS WHERE ProductLine = ?", (row['ProductLine'],))
    product_id = cursor.fetchone()[0] # Fetch the result
    product_ids.append(product_id) # Append the ProducthID to product_ids list.

# Add the new BranchID and ProductID columns to the DataFrame
df['BranchID'] = branch_ids
df['ProductID'] = product_ids

# Create an orders DataFrame by dropping duplicates for adding to the ORDERS table.
orders_df = df[['OrderID', 'BranchID', 'ProductID', 'Quantity', 'UnitPrice', 'OrderDate', 'PaymentMethod', 'CustomerType', 'Gender']].drop_duplicates()

# Add the orders_df to the ORDERS table
orders_df.to_sql('ORDERS', conn, if_exists='append', index=False)


1000

In [232]:
#2

# Select the top 3 ProductLines with the highest total quantity sold from the ORDERS table, joining with the PRODUCTS table.
cursor.execute('''SELECT ProductLine, SUM(Quantity) AS TotalQuantity
FROM ORDERS, PRODUCTS
WHERE ORDERS.ProductID = PRODUCTS.ProductID
GROUP BY ProductLine
ORDER BY TotalQuantity DESC
LIMIT 3;
''')

# Fetch all results from the cursor
results = cursor.fetchall()
columns = ['ProductLine', 'TotalQuantity']

# Create a DataFrame with specified column names
df_results = pd.DataFrame(results, columns = columns)

df_results

ProductLine  TotalQuantity
0  Electronic accessories            971
1      Food and beverages            952
2       Sports and travel            920

In [234]:
#3
# Calculate total revenue per branch and product line
cursor.execute('''SELECT BRANCHES.Branch, PRODUCTS.ProductLine, SUM(ORDERS.UnitPrice * ORDERS.Quantity) AS TotalRevenue
FROM ORDERS, BRANCHES, PRODUCTS
WHERE ORDERS.ProductID = PRODUCTS.ProductID AND ORDERS.BranchID = BRANCHES.BranchID
GROUP BY Branch, ProductLine
''')

# Fetch all results from the cursor
results = cursor.fetchall()
columns = ['Branch', 'ProductLine', 'TotalRevenue']

# Create a DataFrame with specified column names
df_results = pd.DataFrame(results, columns = columns)

df_results

Branch             ProductLine  TotalRevenue
0       A  Electronic accessories      17444.87
1       A     Fashion accessories      15554.77
2       A      Food and beverages      16345.81
3       A       Health and beauty      11997.86
4       A      Home and lifestyle      21349.71
5       A       Sports and travel      18450.19
6       B  Electronic accessories      16239.47
7       B     Fashion accessories      15631.73
8       B      Food and beverages      14490.37
9       B       Health and beauty      19029.20
10      B      Home and lifestyle      16713.49
11      B       Sports and travel      19036.38
12      C  Electronic accessories      18065.69
13      C     Fashion accessories      20533.40
14      C      Food and beverages      22635.10
15      C       Health and beauty      15824.12
16      C      Home and lifestyle      13233.86
17      C       Sports and travel      15011.36

In [236]:
#4
# Calculate total sales per branch
cursor.execute('''SELECT BRANCHES.Branch, SUM(ORDERS.UnitPrice * ORDERS.Quantity) AS TotalSales
FROM ORDERS, BRANCHES
WHERE ORDERS.BranchID = BRANCHES.BranchID
GROUP BY Branch
''')

# Fetch all results from the cursor
results = cursor.fetchall()

columns = ['Branch', 'TotalSales']

# Create a DataFrame with specified column names
df_results = pd.DataFrame(results, columns = columns)

df_results

Branch  TotalSales
0      A   101143.21
1      B   101140.64
2      C   105303.53

In [238]:
#5
# Count the number of orders grouped by gender
cursor.execute('''SELECT Gender, COUNT(*)
FROM ORDERS
GROUP BY Gender
''')

# Fetch all results from the cursor
results = cursor.fetchall()

columns = ['Gender', 'NumberOfOrders']

# Create a DataFrame with specified column names
df_results = pd.DataFrame(results, columns = columns)

df_results

Gender  NumberOfOrders
0  Female             501
1    Male             499

In [242]:
#6
# Count the number of orders for each branch and product line
cursor.execute('''SELECT BRANCHES.Branch, PRODUCTS.ProductLine, COUNT(*)
FROM ORDERS, PRODUCTS, BRANCHES
WHERE ORDERS.ProductID = PRODUCTS.ProductID AND ORDERS.BranchID = BRANCHES.BranchID
GROUP BY BRANCHES.Branch, PRODUCTS.ProductLine
''')

# Fetch all results from the cursor
results = cursor.fetchall()

columns = ['Branch', 'ProductLine', 'NumberOfOrders']

# Create a DataFrame with specified column names
df_results = pd.DataFrame(results, columns = columns)

df_results

Branch             ProductLine  NumberOfOrders
0       A  Electronic accessories              60
1       A     Fashion accessories              51
2       A      Food and beverages              58
3       A       Health and beauty              47
4       A      Home and lifestyle              65
5       A       Sports and travel              59
6       B  Electronic accessories              55
7       B     Fashion accessories              62
8       B      Food and beverages              50
9       B       Health and beauty              53
10      B      Home and lifestyle              50
11      B       Sports and travel              62
12      C  Electronic accessories              55
13      C     Fashion accessories              65
14      C      Food and beverages              66
15      C       Health and beauty              52
16      C      Home and lifestyle              45
17      C       Sports and travel              45

In [249]:
#7
# The ORDERS and PRODUCTS tables are joined, sorted in descending order by quantity, and the `LIMIT 1` clause is used to identify the product line with the highest quantity.
cursor.execute('''SELECT PRODUCTS.ProductLine, ORDERS.Quantity
FROM ORDERS, PRODUCTS
WHERE ORDERS.ProductID = PRODUCTS.ProductID
ORDER BY Quantity DESC
LIMIT 1
''')

# Fetch all results from the cursor
results = cursor.fetchall()

columns = ['ProductLine', 'Quantity']

# Create a DataFrame with specified column names
df_results = pd.DataFrame(results, columns = columns)

df_results


ProductLine  Quantity
0  Home and lifestyle        10

In [250]:
#8
# It retrieves and displays any duplicate OrderID entries in the ORDERS table that appear more than once.
cursor.execute('''SELECT OrderID
FROM ORDERS
GROUP BY OrderID
Having COUNT(*) > 1
''')
results = cursor.fetchall()
print(results)

[]


In [251]:
# Save changes to the database and close the connection
conn.commit()
conn.close()